cpp/src/arrow/python/common.cc

```diff
@@ -122,17 +140,29 @@ Status ConvertPyError(StatusCode code) {
       code = StatusCode::NotImplemented;
     }
   }
+
+  // We consume the python error here so don't construct with a detail.
   return Status(code, message);
 }
 
 Status PassPyError() {
   if (PyErr_Occurred()) {
     // Do not call PyErr_Clear, the assumption is that someone further
     // up the call stack will want to deal with the Python error.
-    return Status(StatusCode::PythonError, "");
+    auto detail = std::make_shared<PythonErrorDetail>();
+    return Status(StatusCode::UnknownError, "", detail);
   }
   return Status::OK();
 }
 
+bool IsPythonError(const Status& status) {
+  if (status.ok()) {
+    return false;
+  }
+  auto detail = status.detail();
+  bool result = detail != nullptr && detail->type_id() == kErrorDetailTypeId;
```

Comment:
Do you think we should use `strcmp` instead? Or is it on purpose?

Reply:
This is on purpose as it should be cheap and I don't think we want to allow "impersonation".  it might be better to type-id be a void* or I can go back to using RTTI (check for nullptr after casting to expected type).   Thoughts?

Reply:
Pointer comparison is fine with me. However, I think we should also use more standardized names, e.g. the fully-qualified class name as I've done in Python (`arrow::py::PythonErrorDetail`).

I don't know if we support building Arrow without RTTI. There are some `dynamic_cast`s in the codebase. See especially `InheritsFrom` in `src/arrow/compute/expression.h`.